### ISSUE TITLE
Failing mpi tests with MPI + OpenMP

### ISSUE THREAD

**Issue by bgranzow** [SNLComputation/omega_h#288]
Compiling Omega_h on a cee machine with the options

```
set(Omega_h_USE_MPI "ON")
set(Omega_h_USE_KokkosCore "OFF")
set(Omega_h_USE_OpenMP "ON")
set(Omega_h_USE_CUDA "OFF")
set(Omega_h_USE_ZLIB "ON")
set(Omega_h_USE_libMeshb "OFF")
set(Omega_h_USE_EGADS "OFF")
set(Omega_h_USE_SEACASExodus "OFF")
set(Omega_h_USE_Teuchos "OFF")
set(Omega_h_USE_DOLFIN "OFF")
set(Omega_h_USE_dwarf "OFF")
set(Omega_h_CHECK_BOUNDS "OFF")
set(Omega_h_THROW "OFF")
set(Omega_h_USE_CUDA_AWARE_MPI "OFF")
set(Omega_h_VERSION_MAJOR "9")
set(Omega_h_VERSION_MINOR "25")
set(Omega_h_VERSION_PATCH "3")
set(Omega_h_SEMVER "9.25.3-sha.f9c37cb+10101000000000")
set(Omega_h_COMMIT "f9c37cba2467327009c5f390648059d375ce12d9")
set(Omega_h_CXX_FLAGS " -O3 -g -Werror -Wall -Wextra -Wdouble-promotion -Wshadow -Wformat=2 -Wduplicated-cond -Wnull-dereference -Wlogical-op -Wrestrict -Wduplicated-branches --std=c++11  -fopenmp -Wno-subobject-linkage ")
set(Omega_h_CMAKE_ARGS "-DBUILD_TESTING:BOOL=\"ON\" -DCMAKE_INSTALL_PREFIX:PATH=\"/scratch/bngranz/nightly/jughead-openmp/install/omega_h\" -DOmega_h_USE_MPI:BOOL=\"ON\" -DOmega_h_USE_OpenMP:BOOL=\"ON\" -DOmega_h_USE_CUDA:BOOL=\"OFF\" -DOmega_h_ARCH:STRING=\"\"")
```

results in a CTest failure in the test `mpi_tests` with the error message:

```
assertion owners.ranks == Read<I32>({1, 0, 1}) failed at /scratch/bngranz/nightly/jughead-openmp/src/omega_h/src/mpi_tests.cpp +189
assertion owners.ranks == Read<I32>({0, 0, 0}) failed at /scratch/bngranz/nightly/jughead-openmp/src/omega_h/src/mpi_tests.cpp +186
```

What's really unfortunate is that this behavior is sporadic, and sometimes the test passes.

We're using `openmpi` version 1.10.1. I recall some mention of one of the openmpi versions not being compatible. Is this it?

**Comment by ibaned** [comment]
Can you point me to the `archie`-style repo and `archie_config` for this offline?

**Comment by bgranzow** [comment]
Tracking progress for this issue:

The point of failure appears to be in [resolve_derived_copies](https://github.com/SNLComputation/omega_h/blob/master/src/Omega_h_build.cpp#L169).

The unit test for this function fails sporadically with `mpirun -n 2 ./mpi_tests` when all other unit tests in in `mpi_tests` are commented out.

After printing every variable in the method `resolve_derived_copies`, the first point of divergence between a failing run and successful run of this unit test is in the variable [sev2vg](https://github.com/SNLComputation/omega_h/blob/master/src/Omega_h_build.cpp#L183).

The inputs to the `dist.exch` function are identical but return different results. Next step will be coding a test with the fixed inputs from the `mpi_tests` executable to see if the behavior is reproduced.

**Comment by bgranzow** [comment]
Code below was used to confirm non-deterministic behavior of `dist.exch`.

This behavior even occurs with OMP_NUM_THREADS=1.

@ibaned is the output of `dist.exch` expected to be deterministic?


```
  Remotes e2ov;
  if (comm->rank() == 0) {
    e2ov.ranks = LOs({0, 0, 0});                                           
    e2ov.idxs = LOs({0, 1, 0});
  } else {
    e2ov.ranks = LOs({0, 1, 0});                                           
    e2ov.idxs = LOs({1, 0, 1});                                            
  }

  LOs ev2vg_canon;
  if (comm->rank() == 0) {
    ev2vg_canon = LOs({0, 1, 1, 3, 0, 3});                                 
  } else {
    ev2vg_canon = LOs({1, 2, 2, 3, 1, 3});                                 
  }

  auto in_dist = Dist(comm, e2ov, 2);                                      
  auto sev2vg = in_dist.exch(ev2vg_canon, 2);                              

  if (comm->rank() == 0) {
    std::cout << sev2vg << std::endl;                                      
  }
  comm->barrier();    

  if (comm->rank() == 1) {
    std::cout << sev2vg << std::endl;                                      
  }
  comm->barrier();  
```

**Comment by ibaned** [comment]
It should be, there is some kind of bug in it for sure. Can you add this to `mpi_tests` so it will always get tested from now on? Then I can pick up this test.

**Comment by ibaned** [comment]
Nevermind, thats not supposed to be deterministic. Its the ordering of edges around a vertex, but its temporary for communication. Only the permanent one is ordered.

**Comment by ibaned** [comment]
The ownership is definitely dependent on this non-deterministic order though. Looks like a genuine oversight, luckily this algorithm is only used when building meshes from raw parallel connectivity, so it shouldn't impact most existing users. Working on a fix now...

**Comment by bgranzow** [comment]
Thank you. I've been busy all morning, but I can also work on this more if you need anything.

**Comment by ibaned** [comment]
I think I know what to do here... if you want to trade, quads in LGRTK are doing some weird things. I pushed some supposed fixes, but shear motion still looks strange to me.

**Comment by ibaned** [comment]
I think this is fixed in `develop` as of c97444771aa70708bb583c3e6e2e0af864b6d37f

Please let me know if you still see any issues

**Comment by bgranzow** [comment]
Yea, I ran this 200 times with `mpirun -n 4 ./mpi_tests` and OMP_NUM_THREADS=4 with no problems. Seems good. Thanks a bunch.
